### PR TITLE
Kylin Network Application

### PR DESCRIPTION
**Pull request by kylin-dev**
# Grant Application Checklist

- [x] The [application-template.md](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( "project_name.md") and updated.
- [x] A BTC address for the payment of the milestones is provided inside the application.  
- [x] The software of the project will be released under the Apache license version 2.0 as specified in the [terms and conditions](https://gist.github.com/Noc2/75bc58e8ce9b5d419ff883b0cf2b8c19).
- [x] The total funding amount of the project is below CHF 30k at the time of submission.

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. It looks really interesting. Could you define the deliveries 1, 2 and 3 a little bit more detailed in the table. These are basically the “requirements” of the contract. 
E.g. will Kylin Network Oracle Node Repo be integrated as a pallet. What are the functions of each delivery? You can also remove the part “connected to the Polkadot network as a parachain”, since this isn’t a requirement we can enforce with the grants program.

**Comment by kylin-dev** [comment]
> Thanks for the application. It looks really interesting. Could you define the deliveries 1, 2 and 3 a little bit more detailed in the table. These are basically the “requirements” of the contract.
> E.g. will Kylin Network Oracle Node Repo be integrated as a pallet. What are the functions of each delivery? You can also remove the part “connected to the Polkadot network as a parachain”, since this isn’t a requirement we can enforce with the grants program.

@Noc2 

- We updated the deliverable 1, 2 and 3 and added more details.
- The deliverables will contain 4 repos and 1 Docker Image.

**Comment by Noc2** [comment]
Hi. Thanks for the update. In general it looks good. I just have two more question: 

- What programming language/framework are you planning to use for the interface? Could you specify this as part of the deliveries. 
- Additionally, I saw that you also shared the exact same proposal with the treasury: https://polkadot.polkassembly.io/post/133 I suggest to focus for now only on one application, either open grants or treasury, because we won’t fund it in case the treasury funds the application.

**Comment by kylin-dev** [comment]
> Hi. Thanks for the update. In general it looks good. I just have two more question:
> 
> * What programming language/framework are you planning to use for the interface? Could you specify this as part of the deliveries.
> * Additionally, I saw that you also shared the exact same proposal with the treasury: https://polkadot.polkassembly.io/post/133 I suggest to focus for now only on one application, either open grants or treasury, because we won’t fund it in case the treasury funds the application.

@Noc2

**Comment by kylin-dev** [comment]
> Hi. Thanks for the update. In general it looks good. I just have two more question:
> 
> * What programming language/framework are you planning to use for the interface? Could you specify this as part of the deliveries.
> * Additionally, I saw that you also shared the exact same proposal with the treasury: https://polkadot.polkassembly.io/post/133 I suggest to focus for now only on one application, either open grants or treasury, because we won’t fund it in case the treasury funds the application.

@Noc2 

Sorry, we did not know that these two applications have conflict. We will cancel the application of treasury funds and only focus on the open grants application.

**Comment by kylin-dev** [comment]
> Hi. Thanks for the update. In general it looks good. I just have two more question:
> 
> * What programming language/framework are you planning to use for the interface? Could you specify this as part of the deliveries.
> * Additionally, I saw that you also shared the exact same proposal with the treasury: https://polkadot.polkassembly.io/post/133 I suggest to focus for now only on one application, either open grants or treasury, because we won’t fund it in case the treasury funds the application.

@Noc2 

- Kylin Network Oracle Node Module Repo will be written in Rust as it is built on top of Substrate as a customized module.
- Kylin Network Datasource Sample Repo  will contain two parts: the Java data retriever to access APIs from exchanges and the NodeJS datasource to handle data feeding.

**Comment by Noc2** [comment]
Thanks! I have two more questions: Are you planning to implement the random selections as part of the PoC (you could take a look at the following [project](https://github.com/Cardinal-Cryptography/substrate/tree/randomness-beacon-m1))? And are you aware of the following [pallet](https://github.com/open-web3-stack/open-runtime-module-library/tree/master/oracle)?

**Comment by kylin-dev** [comment]
> Thanks! I have two more questions: Are you planning to implement the random selections as part of the PoC (you could take a look at the following [project](https://github.com/Cardinal-Cryptography/substrate/tree/randomness-beacon-m1))? And are you aware of the following [pallet](https://github.com/open-web3-stack/open-runtime-module-library/tree/master/oracle)?

@Noc2 

Thanks for your questions.

The random selections is not part of PoC currently as planned, but it will be implemented in the future based on what Substrate providing. When the Randomness Beacon module is released, it could be the cornerstone of our random selections.

The pallet referred in your question is what we investigated before, it's the guide to implement general oracle, but currently our solution is mainly focused on Off-chain Worker which will bring more opportunities for the oracle market.

**Review by Noc2 — APPROVED** [review]
Looks good to me.

**Review by Noc2 — DISMISSED** [review]

**Comment by mmagician** [comment]
@kylin-dev Thanks for the application, the idea sounds interesting. I have a few points that perhaps you could help clarify:

Do I understand correctly that the Arbitrator Node is not part of this grant?
If so, how do you plan to integrate this in the future version of your solution? Is one node going to contain multiple pallets, i.e. Oracle & Arbitrator, or would they be running on separate nodes? Can anyone choose to be an Arbitrator? If so, what's to stop me from running an Arbitrator that colludes with my Oracle (basically the same entity behind both)?

Regarding the Data Warehouse - I didn't quite get whether data will live on-chain or off-chain? How much of historical data will there be stored (and who determines this)? If historical data is placed on-chain, I am not sure just how scalable this would be, as it would grow linearly with the number of data sources and with time.

Thanks for your answers!

File: applications/kylin_network.md [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+* **Decentralized Casinos and Games:** On-chain decentralized casinos and games often require safe and reliable random injection. Kylin Network random number engine provides unpredictable and verifiable random number generation.
+* **Blockchain Computing Market:** Commercial computing such as machine learning training models and 3D rendering needs to complete a variety of complex computing tasks. The off-chain computing market provides verifiable and unlimited off-chain computing capabilities.
+
+#### The Purpose of this Grant
+
+In terms of the system architecture of Kylin Network, the purpose of this grant is to verify features with limited users and launch the testnet for production of concepts. Besides, the implementation of off-chain workers of Substrate 2.0 Framework will be built and validated in a lean way. As we mentioned above, the whole project has four main components, the development roadmap will take at least one year to deliver the platform. At that time, we believe Kylin Network will empower the data economy effectively.
+
+### Ecosystem Fits
+
+There are some existing projects which can support and work with the Kylin Network.
```
[the remark this answers is not in the document]
Comment (kylin-dev) [reply applications/kylin_network.md:130]: We have reached out the co-founders of Plasm and Moonbeam for further collaboration. We are reaching out Ocean and The Graph recently. For rigorous concern, we can remove the projects without confirmation from the ecosystem fits .

**Comment by kylin-dev** [comment]
> @kylin-dev Thanks for the application, the idea sounds interesting. I have a few points that perhaps you could help clarify:
> 
> Do I understand correctly that the Arbitrator Node is not part of this grant?
> If so, how do you plan to integrate this in the future version of your solution? Is one node going to contain multiple pallets, i.e. Oracle & Arbitrator, or would they be running on separate nodes? Can anyone choose to be an Arbitrator? If so, what's to stop me from running an Arbitrator that colludes with my Oracle (basically the same entity behind both)?
> 
> Regarding the Data Warehouse - I didn't quite get whether data will live on-chain or off-chain? How much of historical data will there be stored (and who determines this)? If historical data is placed on-chain, I am not sure just how scalable this would be, as it would grow linearly with the number of data sources and with time.
> 
> Thanks for your answers!

@mmagician 

Thank you for your questions. 

The Arbitrator Node will be included in the future version, and it will be running on separate nodes. Anyone can choose to stake assets to be Arbitrator. Each group selection period there will be only one group of Oracle Nodes (Miner) selected according to randomness grouping algorithm in which the randomness is generated based on the art-of-state feature [randomness-beacon](https://github.com/Cardinal-Cryptography/substrate/tree/randomness-beacon-m1), while the very one group of Arbitrators will be selected in the same way. A random group of Arbitrators will try to validate a random group of Oracle Nodes. 

In the first stage, only real-time data queries will be processed and the whole result will live on-chain if the query is correctly satisfied. In the future version all the data will be stored off-chain in a decentralized storage network such as IPFS, only the queried data will live on-chain without duplicate copies. Later the storage of data can be scaled by sharding, different datasources would have different shards.

**Comment by mmagician** [comment]
@kylin-dev Ok, thanks for the explanation.

> it will be running on separate nodes

So if I, as a user, make a new request, who is the recipient of my query? Is it the Oracle, the Arbitrator or yet another party? 
And if I understood correctly, once my request has been processed, the result is (for now) put directly on the blockchain? How would I know which piece of data was the one requested by me?

File: applications/kylin_network.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+As Kylin Network will provide real-world data to on-chain transactions, security will be placed as a top priority since cyber security issues are quickly becoming a regular occurrence and exposing confidential business information. 
+
+We will employ a novel approach to user identity management in blockchain services. We implement identity-based end-to-end security which extends from the blockchain client to the blockchain fabric. This approach allows for nodes-based network segmentation and traffic separation, which enables multiple entities to securely share the same blockchain infrastructure, reduces the risk of DDoS attacks, and enables automated regulatory compliance audits. The solution is based on Kylin Transport Access Control (KTAC) technologies, implemented using software application library endpoints. This approach can easily be generalized to protect many different types of commercial applications. KTAC features include permission control, confidentiality, un-linkable identity privacy for blockchain participants, a modular and easily auditable consensus protocol, and improved scalability. KTAC extends the blockchain in several important ways.
```
[the remark this answers is not in the document]
Comment (kylin-dev) [reply applications/kylin_network.md:106]: Unfortunately, there is no spec for KTAC yet, and it's for the future version. The general idea of this protocol is to handle the case of dedicated datasource(s) for some data consumers, and it's used together with KRTP to protect datasource from unauthorized access. The draft process would be that the data consumer sends a query request with the signed signature and the related customized Oracle (running with Oracle Node) will verify the signature with access control info for the datasource. If the signature is valid for this datasource, then the transportation for data would begin as normal.

File: applications/kylin_network.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+We will employ the Kylin Real-time Transport Protocol (KRTP), a profile of the Real-time Transport Protocol (RTP), which provides confidentiality, message and data authentication, and replay protection to the RTP traffic and to the control traffic for RTP, the Real-time Transport Control Protocol (RTCP). KRTP provides a framework for encryption and data authentication of RTP and RTCP streams. KRTP defines a set of cryptographic transforms, and it allows new transforms to be introduced in the future. With appropriate key management, KRTP is secure for unicast and multicast RTP applications.
```
[the remark this answers is not in the document]
Comment (kylin-dev) [reply applications/kylin_network.md:108]: KTAC and KRTP are for the future version, they are still in early design stage, and no spec yet. As mentioned in the previous answer of KTAC, KRTP is the used together with KTAC to make sure the security and privacy of data passing through Kylin Network. As we are trying to provide general Oracle Market, and the encryption algorithm is public, KTAC just controls the access but not able to protect the data during transportation, that means we need more work. We initially thought RTP is the mature and widely used solution for the real-time data transportation currently, implement dynamic encryption over RTP would archive the goal to protect data privacy. Maybe it's not a proper direction to solve the problem. We are still actively working on this problem to find a proper solution to bring added value for the future version.

**Comment by kylin-dev** [comment]
> @kylin-dev Ok, thanks for the explanation.
> 
> > it will be running on separate nodes
> 
> So if I, as a user, make a new request, who is the recipient of my query? Is it the Oracle, the Arbitrator or yet another party?
> And if I understood correctly, once my request has been processed, the result is (for now) put directly on the blockchain? How would I know which piece of data was the one requested by me?

@mmagician 

Thank you for your comments. 

It's the Oracle which handles the query request in current design. The result will be validated across the group of Oracle Nodes to make sure the result from the datasource is honestly fetched. The query would have identifier field which is derived from the key-pairs of user and data source id, and this identifier would be in the result data, then the SDK for user side will extract the very piece of data.

**Comment by mmagician** [comment]
@kylin-dev Thanks for the explanation.

So if I understand correctly, your milestone 1 will provide two substrate pallets: Oracle Node & Data Feeding? Or would these be part of the same pallet?

Regardless, could you please include the technical specifications for each pallet that you propose, to include the following details:
- List the publicly exposed methods
- For each method, specify (to the best of present knowledge):
    - Method signature (incl. parameters with their types, return type)
    - Short description (code template)
- Runtime Storage defined by your module

Also, you have the diagrams for your final proposed solution, which includes elements such as the Arbitrator Node that are not part of M1. Could you also supply a use case diagram demonstrating how system components & external users interact with one another, but targeting the state of your solution at the point of completion of milestone 1?

Why exactly do you need the `Datasource Sample Repo` and what is its functionality? Am I right in understanding that the `Data Feeding` module will interact with it? Couldn't the off-chain worker reach out directly to the off-chain source via an http request? I just fail to grasp the purpose of this sample repo.

Many thanks!

File: applications/kylin_network.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (laboon) [reply applications/kylin_network.md:108]: I'm still not sure why you would care if the encryption algorithm is public?  And why you couldn't use something like TLS?  Another protocol seems really off the path of what you would need for oracles if the data is already on the network and being communicated through nodes.

[the remark this answers is not in the document]
Comment (kylin-dev) [reply applications/kylin_network.md:108]: That's a good point! The initial idea is to protect the end-to-end transportation, TLS is also an optional solution. We will investigate more tech solutions to achieve that goal in the future version. Thank you for your comments.

**Comment by kylin-dev** [comment]
> @kylin-dev Thanks for the explanation.
> 
> So if I understand correctly, your milestone 1 will provide two substrate pallets: Oracle Node & Data Feeding? Or would these be part of the same pallet?
> 
> Regardless, could you please include the technical specifications for each pallet that you propose, to include the following details:
> 
> * List the publicly exposed methods
> * For each method, specify (to the best of present knowledge):
>   
>   * Method signature (incl. parameters with their types, return type)
>   * Short description (code template)
> * Runtime Storage defined by your module
> 
> Also, you have the diagrams for your final proposed solution, which includes elements such as the Arbitrator Node that are not part of M1. Could you also supply a use case diagram demonstrating how system components & external users interact with one another, but targeting the state of your solution at the point of completion of milestone 1?
> 
> Why exactly do you need the `Datasource Sample Repo` and what is its functionality? Am I right in understanding that the `Data Feeding` module will interact with it? Couldn't the off-chain worker reach out directly to the off-chain source via an http request? I just fail to grasp the purpose of this sample repo.
> 
> Many thanks!

@mmagician 

> So if I understand correctly, your milestone 1 will provide two substrate pallets: Oracle Node & Data Feeding? Or would these be part of the same pallet?

The original idea is to have separate substrate pallets: Oracle Node and Data Feeding.

> Regardless, could you please include the technical specifications for each pallet that you propose, to include the following details:
> 
> * List the publicly exposed methods
> * For each method, specify (to the best of present knowledge):
>   
>   * Method signature (incl. parameters with their types, return type)
>   * Short description (code template)
> * Runtime Storage defined by your module

> Also, you have the diagrams for your final proposed solution, which includes elements such as the Arbitrator Node that are not part of M1. Could you also supply a use case diagram demonstrating how system components & external users interact with one another, but targeting the state of your solution at the point of completion of milestone 1?

In terms of technical specifications and use-case diagram, we will update the related information in application.

> Why exactly do you need the `Datasource Sample Repo` and what is its functionality? Am I right in understanding that the `Data Feeding` module will interact with it? Couldn't the off-chain worker reach out directly to the off-chain source via an http request? I just fail to grasp the purpose of this sample repo.

In our first version, we would like to use market data from exchanges. To access the API of exchanges, the API key and API secret are required. Each exchange will have its own way to secure the API access, so it's not convenient to access those APIs directly with the off-chain worker, and we will provide a component to handle such API access related stuff.

**Comment by mmagician** [comment]
@kylin-dev Great! I'll be waiting to see the specs then 👍 

> Each exchange will have its own way to secure the API access

But it's not in the scope of your project (or is it?) to make a connector to each of the exchanges & host those in a separate repp, right? This is purely for the demonstration purposes, if I understand you correctly? 
Anyway, how will it exactly work, then? Will you host an intermediary service which accepts easy-to-code requests from off-chain workers & then in that service, create the full request to the data provider, with added authentication?

**Comment by kylin-dev** [comment]
> @kylin-dev Great! I'll be waiting to see the specs then 👍
> 
> > Each exchange will have its own way to secure the API access
> 
> But it's not in the scope of your project (or is it?) to make a connector to each of the exchanges & host those in a separate repp, right? This is purely for the demonstration purposes, if I understand you correctly?
> Anyway, how will it exactly work, then? Will you host an intermediary service which accepts easy-to-code requests from off-chain workers & then in that service, create the full request to the data provider, with added authentication?

@mmagician 

To be precise, it's not in the scope of Kylin Network project, but for demonstration purposes to let anyone can REALLY use the deliveries. We will not host any service but let anyone who wanna use the demonstration to run the service by themselves. The process is like what you mentioned. We will provide the Datasource sample repo, the hacker use that repo and run the service by only fill in the API key & API secret. The service will accept easy-to-code requests from off-chain workers, and the service will handle the rest.

**Comment by kylin-dev** [comment]
> @kylin-dev Thanks for the explanation.
> 
> So if I understand correctly, your milestone 1 will provide two substrate pallets: Oracle Node & Data Feeding? Or would these be part of the same pallet?
> 
> Regardless, could you please include the technical specifications for each pallet that you propose, to include the following details:
> 
> * List the publicly exposed methods
> * For each method, specify (to the best of present knowledge):
>   
>   * Method signature (incl. parameters with their types, return type)
>   * Short description (code template)
> * Runtime Storage defined by your module
> 
> Also, you have the diagrams for your final proposed solution, which includes elements such as the Arbitrator Node that are not part of M1. Could you also supply a use case diagram demonstrating how system components & external users interact with one another, but targeting the state of your solution at the point of completion of milestone 1?
> 
> Why exactly do you need the `Datasource Sample Repo` and what is its functionality? Am I right in understanding that the `Data Feeding` module will interact with it? Couldn't the off-chain worker reach out directly to the off-chain source via an http request? I just fail to grasp the purpose of this sample repo.
> 
> Many thanks!

@mmagician 

We updated the application proposal and add 'Pallet Design and Implements' into the Polkadot/Substrate Integration part. It shows the  technical specifications and use-case diagram.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Could you re-submit the application and close the current one? It’s difficult to accept it in the current stage (see cla assistant check, 322 commits).

**Comment by kylin-dev** [comment]
> Could you re-submit the application and close the current one? It’s difficult to accept it in the current stage (see cla assistant check, 322 commits).

Sure, no problem

**Pull request closed without merging**